apps/frontend/src/app/5_pages/ConvertPage/hooks/useHandleConversion.ts

```diff
@@ -137,34 +117,121 @@ export const useHandleConversion = (
     return transactions;
   }, [account, getMassetManager, signer, sourceToken, weiAmount, onComplete]);
 
-  const depositTokens = useCallback(async () => {
-    if (!signer || destinationToken !== SupportedTokens.dllr) {
+  const handleSubmit = useCallback(async () => {
+    if (!route || !signer) {
       return;
     }
-    const transactions = await getDepositTokenTransactions();
+
+    const [sourceTokenDetails, destinationTokenDetails] = await Promise.all([
+      getTokenDetails(sourceToken, defaultChainId),
+      getTokenDetails(destinationToken, defaultChainId),
+    ]);
+
+    const approveTxData = await route.approve(
+      sourceTokenDetails.address,
+      destinationTokenDetails.address,
+      weiAmount,
+      account,
+    );
+
+    const transactions: Transaction[] = [];
+
+    if (approveTxData && approveTxData.to && approveTxData.data) {
+      transactions.push({
+        title: t(translations.convertPage.txDialog.approve, {
+          asset: sourceToken.toUpperCase(),
+        }),
+        request: {
+          type: TransactionType.signTransactionData,
+          signer: signer,
+          to: approveTxData.to,
+          data: approveTxData.data,
+          gasLimit: GAS_LIMIT.CONVERT,
+        },
+        onComplete,
+      });
+    }
+
+    const permitTxData = await route.permit(
+      sourceTokenDetails.address,
+      destinationTokenDetails.address,
+      weiAmount,
+      account,
+    );
+
+    if (permitTxData) {
+      transactions.push(
+        await preparePermitTransaction({
+          token: sourceTokenDetails.symbol,
+          signer,
+          spender: permitTxData.spender,
+          value: permitTxData.value?.toString(),
+          deadline: permitTxData.deadline,
+          nonce: permitTxData.nonce,
+        }),
+      );
+    }
+
+    const txData = await route.swap(
+      sourceTokenDetails.address,
+      destinationTokenDetails.address,
+      weiAmount,
+      account,
+    );
```

Comment:
Put UNSIGNED_PERMIT to options if permit is expected, otherwise you will get an exception here

```suggestion
    const txData = await route.swap(
      sourceTokenDetails.address,
      destinationTokenDetails.address,
      weiAmount,
      account,
      { permit: permitTxData ? UNSIGNED_PERMIT : undefined }
    );
```